File: windows/BlueLink.App/Domain/FileBatch.cs
namespace BlueLink.Domain;

public enum FileBatchAction { Retry, Cancel, DeleteRecords, Copy }
public enum FileBatchOutcome { Submitted, Skipped, Failed }
public sealed record FileBatchEntry(Guid Id, string Name, FileBatchOutcome Outcome, string? Reason = null);
public sealed record FileBatchCheck(Guid Id, string Name, bool Eligible, string? Reason = null);

public static class FileBatchPolicy
{
    public const int MaximumSelection = 100;
    public static FileBatchCheck Check(Guid id, TransferItem? item, FileBatchAction action, bool online)
    {
        string? reason = item is null ? "记录已不存在" : action switch {
            FileBatchAction.Retry when !item.Outgoing || !item.IsRetryableTerminal => "当前状态不支持重试",
            FileBatchAction.Retry when !online => "设备当前未连接",
            FileBatchAction.Retry when !item.CanRetry => "原文件不可读取",
            FileBatchAction.Cancel when !item.IsActive => "任务已结束",
            FileBatchAction.Cancel when !online => "设备当前未连接",
            FileBatchAction.DeleteRecords when !item.CanDelete => "请先取消正在进行的任务",
            FileBatchAction.Copy when !item.IsCompleted || !item.CanOpen => "文件尚未完成或不可读取",
            _ => null };
        return new(id,item?.Name ?? id.ToString(),reason is null,reason);
    }
}

public static class FileBatchRunner
{
    public static async Task<IReadOnlyList<FileBatchEntry>> RunAsync(IEnumerable<Guid> selection,
        Func<Guid,Task<FileBatchCheck>> check, Func<Guid,Task> execute, CancellationToken token = default)
    {
        var ids = selection.Distinct().ToArray();
        if(ids.Length > FileBatchPolicy.MaximumSelection) throw new ArgumentOutOfRangeException(nameof(selection));
        var results = new List<FileBatchEntry>();
        foreach(var id in ids)
        {
            token.ThrowIfCancellationRequested();
            var name = id.ToString();
            try
            {
                var current = await check(id); name=current.Name;
                if(!current.Eligible) { results.Add(new(id,name,FileBatchOutcome.Skipped,current.Reason)); continue; }
                await execute(id);
                results.Add(new(id,name,FileBatchOutcome.Submitted));
            }
            catch(OperationCanceledException) when(token.IsCancellationRequested) { throw; }
            catch(Exception error) { results.Add(new(id,name,FileBatchOutcome.Failed,error.Message)); }
        }
        return results;
    }
}

File: android/app/src/main/java/com/bluelink/android/domain/FileBatch.kt
package com.bluelink.android.domain

import java.util.UUID
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.currentCoroutineContext
import kotlinx.coroutines.ensureActive

enum class FileBatchAction { RETRY, CANCEL, DELETE_RECORDS, SHARE }
enum class FileBatchOutcome { SUBMITTED, SKIPPED, FAILED }
enum class FileBatchReason { MISSING, STATE_CHANGED, OFFLINE, SOURCE_UNREADABLE, ACTIVE, FAILED }
data class FileBatchCheck(val id: UUID, val name: String, val eligible: Boolean, val reason: FileBatchReason? = null)
data class FileBatchEntry(val id: UUID, val name: String, val outcome: FileBatchOutcome, val reason: FileBatchReason? = null)

interface FileBatchOperations {
    suspend fun check(ids: List<UUID>, action: FileBatchAction): List<FileBatchCheck>
    suspend fun run(ids: List<UUID>, action: FileBatchAction): List<FileBatchEntry>
}

object FileBatchPolicy {
    const val MAXIMUM_SELECTION = 100
    fun check(id: UUID, item: TransferItem?, action: FileBatchAction, online: Boolean, readable: Boolean): FileBatchCheck {
        val reason = when {
            item == null -> FileBatchReason.MISSING
            action == FileBatchAction.RETRY && (!item.outgoing || item.status !in setOf(TransferStatus.FAILED,TransferStatus.REJECTED,TransferStatus.CANCELED)) -> FileBatchReason.STATE_CHANGED
            (action == FileBatchAction.RETRY || action == FileBatchAction.CANCEL) && !online -> FileBatchReason.OFFLINE
            action == FileBatchAction.RETRY && !readable -> FileBatchReason.SOURCE_UNREADABLE
            action == FileBatchAction.CANCEL && item.status !in HistoryQuery.activeStatuses -> FileBatchReason.STATE_CHANGED
            action == FileBatchAction.DELETE_RECORDS && item.status in HistoryQuery.activeStatuses -> FileBatchReason.ACTIVE
            action == FileBatchAction.SHARE && (item.status != TransferStatus.COMPLETED || !readable) -> FileBatchReason.SOURCE_UNREADABLE
            else -> null
        }
        return FileBatchCheck(id,item?.name ?: id.toString(),reason == null,reason)
    }
}

object FileBatchRunner {
    suspend fun run(selection: List<UUID>, check: suspend (UUID)->FileBatchCheck, execute: suspend (UUID)->Unit): List<FileBatchEntry> {
        val ids=selection.distinct()
        require(ids.size <= FileBatchPolicy.MAXIMUM_SELECTION)
        return ids.map { id ->
            currentCoroutineContext().ensureActive()
            var name=id.toString()
            try {
                val current=check(id); name=current.name
                if(!current.eligible) FileBatchEntry(id,name,FileBatchOutcome.SKIPPED,current.reason)
                else { execute(id); FileBatchEntry(id,name,FileBatchOutcome.SUBMITTED) }
            } catch(cancelled: CancellationException) { currentCoroutineContext().ensureActive(); FileBatchEntry(id,name,FileBatchOutcome.FAILED,FileBatchReason.FAILED) }
            catch(error: Exception) { FileBatchEntry(id,name,FileBatchOutcome.FAILED,FileBatchReason.FAILED) }
        }
    }
}
